ergo/foretold.py

```diff
@@ -0,0 +1,70 @@
+from matplotlib import pyplot as plt
+import torch
+import numpy as np
+import requests
+from ergo.ppl import uniform
+
+
+class ForetoldDistribution:
+    """"Aggregated distribution from a foretold question"""
+
+    def __init__(self, id):
+        """
+            id: measurableId, the second id in the URL for a foretold question
+        """
+        self.id = id
+        self.floatCdf = None
+        # previousAggregate is the most recent aggregated distribution
+        response = requests.post(
+            "https://prediction-backend.herokuapp.com/graphql",
+            json={
+                "variables": {"measurableId": self.id},
+                "query": """query ($measurableId: String!) {
+      measurable(id:$measurableId) {
+        id
+        channelId
+        previousAggregate {
+          value {
+            floatCdf {
+              xs
+              ys
+            }
+          }
+        }
+      }
+    }
+    """,
+            },
+        )
+        j = response.json()
+        try:
+            self.channelId = j["data"]["measurable"]["channelId"]
+            self.floatCdf = j["data"]["measurable"]["previousAggregate"]["value"][
+                "floatCdf"
+            ]
+        except KeyError:
+            raise (f"Error loading distribution {self.id} from Foretold")
+
+    @property
+    def url(self):
+        return f"https://www.foretold.io/c/{self.channelId}/m/{self.id}"
+
+    def sample(self):
+        """Sample from CDF 
+
+        First sample between 0 and 1, find the corresponding bin, then linearly interpolate within the bin"""
+        xs = torch.tensor(self.floatCdf["xs"])
+        ys = torch.tensor(self.floatCdf["ys"])
+        y = uniform()
+        i = np.argmax(ys > y)
+        if i == len(ys) - 1:
```

Comment:
seems like when this condition is met, the x returned will be a bit lower than it should have been?

because in this case we're just taking the last xs[i], when ideally we'd want to add a bit to it, since ys > than the last y[i]

(not sure what to do about this, mostly just checking my understanding)

Reply:
I actually had an off-by-one error in the whole thing. Now it makes more sense, because the first ys should be 0, so we never sample a lower value than 0.